onnxruntime/core/providers/cuda/math/cumsum_impl.cu

```diff
@@ -0,0 +1,160 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/providers/cuda/cu_inc/common.cuh"
+#include "cumsum_impl.h"
+
+namespace onnxruntime {
+namespace cuda {
+
+template <typename T>
+__global__ void _CumSumKernel(
+    const T* input_data,
+    const int64_t input_dim_along_axis,
+    const int64_t input_stride_along_axis,
+    T* output_data,
+    const int64_t output_size,
+    const bool exclusive,
+    const bool reverse) {
+  CALCULATE_ELEMENTWISE_INDEX_OR_EXIT(indices_index, output_size);
+
+  int axis_dim = (indices_index / input_stride_along_axis) % input_dim_along_axis;
+
+  int start = 0;
+  int end = 0;
+
+  if (!reverse && !exclusive) {
+    start = 0;
+    end = axis_dim;
+  
+  } else if (reverse && !exclusive) {
+    start = axis_dim;
+    end = input_dim_along_axis - 1;
+
+  } else if (!reverse && exclusive) {
+    start = 0;
+    end = axis_dim - 1;
+
+  } else { // reverse && exclusive
+    start = axis_dim + 1;
+    end = input_dim_along_axis - 1;
+
+  }
+
+  // count the number of elements to accumulate the sum
+  int count = end - start + 1;
+  if (count <= 0) {
+    output_data[indices_index] = 0;
+    return;  
+  }
+
+  // adjust start index based on the above identified start dim value along the axis of interest
+  int64_t data_index = static_cast<int64_t>(indices_index) + (static_cast<int64_t>(start - axis_dim) * input_stride_along_axis);
+  T sum = 0;
+
+  // keep accumulating values from the start index for 'count' times and skip appropriately 
+  while (count != 0) {
+    sum += input_data[data_index];
+    data_index += input_stride_along_axis;
+    --count;
+  }
```

Comment:
Not sure if the perf would be ideal. Maybe better to use cub::DeviceScan

Reply:
This looks promising but does have some limitations that may not make it conducive for usage in this scenario. From what I see, InclusiveScan/InclusiveSum does not support arbitrary element skips which is something we need here. If we need to get it to accumulate all the elements that need to get summed together in a parallel fashion, we might as well just accumulate the sum along the way and there is no need to call parallel sum/scan.

Reply:
If this is not the bottleneck in your model, it would be OK. Otherwise, we may special case for step=1 to use CUB, and fallback to this or doing transpose before/after if that's faster.

---
In reply to: [357877759](https://github.com/microsoft/onnxruntime/pull/2647#discussion_r357877759) [](ancestors = 357877759)

Reply:
Should we wait until we have a specific model that we are tuning perf for before we add such fine-grained logic of adding transpose before/after the op ?

You are right - it can be special cased for input_stride_along_axis == 1. In that case, we could use InclusiveSum. But I don't quite know if it will be faster than this approach. We may have to profile for the specific model that we are tuning the perf for. Also, we need to allocate a temporary buffer before using the cub approach (this is from the link you shared).

Reply:
I thought that the implementation is driven by the performance need of some model. If this implementation already made this op not being the bottleneck, please go ahead.

Reply:
Thanks.